core/src/main/java/com/linecorp/armeria/client/limit/AbstractConcurrencyLimitingClient.java

```diff
@@ -84,30 +84,43 @@ protected AbstractConcurrencyLimitingClient(Client<I, O> delegate, int maxConcur
      */
     protected AbstractConcurrencyLimitingClient(Client<I, O> delegate,
                                                 int maxConcurrency, long timeout, TimeUnit unit) {
-        super(delegate);
-
-        validateAll(maxConcurrency, timeout, unit);
+        this(delegate, ConcurrencyLimit.builder()
+                                       .maxConcurrency(maxConcurrency)
+                                       .timeoutMillis(timeout)
+                                       .build());
+    }
 
-        if (maxConcurrency == Integer.MAX_VALUE) {
-            this.maxConcurrency = 0;
-        } else {
-            this.maxConcurrency = maxConcurrency;
-        }
-        timeoutMillis = unit.toMillis(timeout);
+    /**
+     * Creates a new instance that decorates the specified {@code delegate} to limit the concurrent number of
+     * active requests to {@code maxConcurrency}.
+     *
+     * @param delegate the delegate {@link Client}
+     * @param concurrencyLimit the concurrency limit config
+     */
+    protected AbstractConcurrencyLimitingClient(Client<I, O> delegate, ConcurrencyLimit concurrencyLimit) {
+        super(delegate);
+        this.concurrencyLimit = concurrencyLimit;
+        numActiveRequests = concurrencyLimit.numActiveRequests();
     }
 
     static void validateAll(int maxConcurrency, long timeout, TimeUnit unit) {
```

Comment:
I think we can now move this method into `ConcurrencyLimitingClient` and make it private. 😉

Reply:
Yeah, so I was trying mostly not to change the way the project was handling the validations keeping the old API the same.

Right now there are multiple places where we are doing validations of the same values.
i.e timeout is validated both in the CuncorrencyLimitBuilder + in the Client.

I am happy to delegate all the validation to happen in the builders which means we don't do the validation in the `newDecorator(...)`  method. WDYT?

Reply:
> I am happy to delegate all the validation to happen in the builders which means we don't do the validation in the newDecorator(...) method. WDYT?

That's a good suggestion. 👍